04-onion-routing.md

```diff
@@ -367,6 +381,249 @@ otherwise meets the amount criterion (eg. some other failure, or
+    the route is used in the right context
+  - MAY add padding data to ensure all `encrypted_data` have the same length
+- MUST communicate the blinded node IDs `B(i)` and `encrypted_data(i)` to the sender
+- MUST communicate the real node ID of the introduction point `N(0)` to the sender
+- MUST communicate the first ephemeral key `E(0)` to the sender
+
+The sender:
+
+- MUST provide the `encrypted_data` field for each node in the blinded route
+  in their respective onion payloads
+- MUST communicate `E(0)` to the introduction point
+
+The introduction point:
+
+- MUST compute:
+  - `ss(0) = SHA256(k(0) * E(0))` (standard ECDH)
+  - `rho(0) = HMAC256("rho", ss(0))`
+  - `E(1) = SHA256(E(0) || ss(0)) * E(0)`
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(0)` and `E(0)`
+  - MUST use the decrypted fields to locate the next node
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
+  - MUST use `E(1)` as the next blinding point
+- MUST forward the onion and communicate the next blinding point to the next node
+
+An intermediate node in the blinded route:
+
+- MUST compute:
+  - `ss(i) = SHA256(k(i) * E(i))` (standard ECDH)
+  - `b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)`
+  - `rho(i) = HMAC256("rho", ss(i))`
+  - `E(i+1) = SHA256(E(i) || ss(i)) * E(i)`
+- MUST use `b(i)` instead of its private key `k(i)` to decrypt the onion. Note
+  that the node may instead tweak the onion ephemeral key with
+  `HMAC256("blinded_node_id", ss(i))` which achieves the same result.
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(i)` and `E(i)`
+  - MUST use the decrypted fields to locate the next node
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
+  - MUST use `E(i+1)` as the next blinding point
+- MUST forward the onion and communicate the next blinding point to the next node
+
+The final recipient:
+
+- MUST compute:
+  - `ss(r) = SHA256(k(r) * E(r))` (standard ECDH)
+  - `b(r) = HMAC256("blinded_node_id", ss(r)) * k(r)`
+  - `rho(r) = HMAC256("rho", ss(r))`
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(r)` and `E(r)`
+- MUST ignore the message if `E(r)` or `path_id` does not match the
+  blinded route it created
+
+#### Rationale
+
+Route blinding is a lightweight technique to provide recipient anonymity.
+It's more flexible than rendezvous because it lets senders choose what data
+they include in each node's onion payload and retry with updated values if
+the route fails.
+
+Use-cases where route blinding is useful include:
+
+- Recipient anonymity when receiving payments
+- Using unannounced channels in invoices without revealing them
+- Forcing a message to go through a specific set of intermediaries that can witness it
+- Providing anonymous reply paths for onion messages (offers, stuckless payments, etc)
```

Comment:
To what extent is there potential here for an amplification attack? Bombarding nodes with invoice requests and directing them to deliver the invoice across 27 hops.

Reply:
That's not really an amplification attack is it? You can more easily do that yourself. But it does make it even harder to pinpoint the attacker.

Reply:
I guess it depends on the definition. It isn't snowball-amplification, but there is a multiplying factor nonetheless. You can indeed do it yourself, but there may be a reason why it is advantageous to shift blame onto someone else.